### ISSUE TITLE
REST API wrong behaviour

### ISSUE THREAD

**Issue by vladbogo** [buildbot/buildbot#5207]
Hi,

I have a question regrading the REST API. We are trying to get a list of the last changes for a particular branch. 

However, using https://buildbot.mariadb.org/api/v2/changes?branch=10.1&order=-changeid&limit=10 does not return the last 10 changes for the 10.1 branch. Is this the expected behaviour? Based on the documentation I would assume that this would return the last 10 changes for branch 10.1. If the limit argument is increased, more changes start to appear.

Things get a lot more confusing when deleting the "order" argument https://buildbot.mariadb.org/api/v2/changes?branch=10.1&limit=10 when the request times out and the buildbot is unresponsive for several minutes. Any thoughts on this? Is there any sort of cache used?

Thanks!

**Comment by vladbogo** [comment]
Any updates?

**Comment by tardyp** [comment]
Indeed, this is a bug, we do not support limit and filtering with the current code.

as per this code, the limit is done here, but only when order is -changeid

https://github.com/buildbot/buildbot/blob/81d693a8e973b51552cd23bfa695c4b2b0a30f2c/master/buildbot/data/changes.py#L85

This should rather be implemented like in the builds endpoint:
https://github.com/buildbot/buildbot/blob/81d693a8e973b51552cd23bfa695c4b2b0a30f2c/master/buildbot/data/builds.py#L161
where resultSpec is passed to the db api

and then used with a generic database query generation
https://github.com/buildbot/buildbot/blob/81d693a8e973b51552cd23bfa695c4b2b0a30f2c/master/buildbot/db/builds.py#L147

would you like to fix this code?

**Comment by vladbogo** [comment]
Thanks for your answer. I will look into it and try to fix it.
